obs2ioda-v2/src/f_c_string_t_mod.f90

```diff
@@ -0,0 +1,54 @@
+module f_c_string_t_mod
+    use iso_c_binding, only : c_loc, c_ptr, c_null_char, c_char, c_f_pointer, c_null_ptr
+    use f_c_string_i_mod, only : strlen
+    implicit none
+
+    type :: f_c_string_t
+        ! Allocatable Fortran string
+        character(len = :), allocatable :: f_string
+        ! Allocatable C-compatible null-terminated string
+        character(len = :, kind = c_char), allocatable :: fc_string
+        ! C pointer to the null-terminated string
+        type(c_ptr) :: c_string
```

Comment:
Is this needed?

Reply:
`fc_string` is necessary, but `c_string` is not—good catch! I verified this by running the `f_c_string` test with `fc_string` defined as a local variable in `to_c`. The test failed when calling `strlen` on the `c_string` variable returned by `to_c`.

Reply:
You can't return a pointer to a local variable, but can you return the array, instead of a pointer to the array?

Usage would look like
```
character(len = :, kind = c_char), allocatable :: my_c_string
character(len = :), allocatable :: f_string

my_c_string = to_c('now is the time')
! ... use the cstring
f_string = to_f(my_c_string)
! when my_c_string goes out of scope it will be deallocated.
! when f_string goes out of scope it will be deallocated
```
`to_f()` would take a `character(len = :, kind = c_char), allocatable` instead of a `type(c_ptr)`

Or am I misunderstanding Fortran allocatable arrays?

Reply:
If this is doable I think it will simplify the implementation of f_c_string_1D_t_mod